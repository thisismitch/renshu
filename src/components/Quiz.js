import React, { Component } from 'react';
import styled from '@emotion/styled';

// UI elements
import { Button, ProgressBar, TextInput } from './ui';
import { ReactComponent as ChevronRight } from '../assets/round-chevron_right-24px.svg';

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  min-height: 100vh;
  background-color: #f5f5f5;
`;

const Content = styled.div`
  margin-bottom: 3em;
  text-align: center;
`;

const Prompt = styled.h1`
  margin: 0 0 2rem 0;
  font-size: 32vmin;
  line-height: 1;
  text-align: center;
`;

class Quiz extends Component {
  static defaultProps = {
    placeholder: 'Enter sound (romaji)',
  };

  constructor(props) {
    super(props);

    const all = [...props.questions];

    this.state = {
      all,
      initialLength: all.length,
      current: all[0],
      answer: '',
      inputPlaceholder: props.placeholder,
    };

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(e) {
    const val = e.target.value.trim().toLowerCase();
    const { current } = this.state;

    if (val === current.sound || val === current.romaji) {
      const remainingCharacters = this.state.all.slice(1);
      return this.setState({
        all: remainingCharacters,
        current: remainingCharacters[0],
        answer: '',
      });
    }

    this.setState({ answer: val });
  }

  render() {
    const {
      all,
      current,
      answer,
      initialLength,
      inputPlaceholder,
    } = this.state;
    const { placeholder } = this.props;
    const progress = ((initialLength - all.length) / initialLength) * 100;

    return (
      <>
        <ProgressBar progress={progress} />
        <Wrapper>
          <Content>
            {current ? (
              <>
                <Prompt>{current.character}</Prompt>
                <TextInput
                  type="text"
                  value={answer}
                  onChange={this.handleChange}
                  placeholder={inputPlaceholder}
                  onBlur={() =>
                    this.setState({ inputPlaceholder: placeholder })
                  }
                  onFocus={() => this.setState({ inputPlaceholder: '' })}
                  autoFocus={true}
                />
              </>
            ) : (
              <>
                <Prompt>すごいよ!</Prompt>
                <Button onClick={() => window.location.reload()}>
                  Start again <ChevronRight />
                </Button>
              </>
            )}
          </Content>
        </Wrapper>
      </>
    );
  }
}

export default Quiz;
