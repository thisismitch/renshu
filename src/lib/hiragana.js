import ROWS from "./rows";

export default [
  { sound: "a", character: "あ", row: ROWS.A },
  { sound: "i", character: "い", row: ROWS.A },
  { sound: "u", character: "う", row: ROWS.A },
  { sound: "e", character: "え", row: ROWS.A },
  { sound: "o", character: "お", row: ROWS.A },

  { sound: "ka", character: "か", row: ROWS.KA },
  { sound: "ki", character: "き", row: ROWS.KA },
  { sound: "ku", character: "く", row: ROWS.KA },
  { sound: "ke", character: "け", row: ROWS.KA },
  { sound: "ko", character: "こ", row: ROWS.KA },

  { sound: "sa", character: "さ", row: ROWS.SA },
  { sound: "shi", character: "し", row: ROWS.SA },
  { sound: "su", character: "す", row: ROWS.SA },
  { sound: "se", character: "せ", row: ROWS.SA },
  { sound: "so", character: "そ", row: ROWS.SA },

  { sound: "ta", character: "た", row: ROWS.TA },
  { sound: "chi", character: "ち", row: ROWS.TA },
  { sound: "tsu", character: "つ", row: ROWS.TA },
  { sound: "te", character: "て", row: ROWS.TA },
  { sound: "to", character: "と", row: ROWS.TA },

  { sound: "na", character: "な", row: ROWS.NA },
  { sound: "ni", character: "に", row: ROWS.NA },
  { sound: "nu", character: "ぬ", row: ROWS.NA },
  { sound: "ne", character: "ね", row: ROWS.NA },
  { sound: "no", character: "の", row: ROWS.NA },

  { sound: "ha", character: "は", row: ROWS.HA },
  { sound: "hi", character: "ひ", row: ROWS.HA },
  { sound: "fu", character: "ふ", row: ROWS.HA },
  { sound: "he", character: "へ", row: ROWS.HA },
  { sound: "ho", character: "ほ", row: ROWS.HA },

  { sound: "ma", character: "ま", row: ROWS.MA },
  { sound: "mi", character: "み", row: ROWS.MA },
  { sound: "mu", character: "む", row: ROWS.MA },
  { sound: "me", character: "め", row: ROWS.MA },
  { sound: "mo", character: "も", row: ROWS.MA },

  { sound: "ra", character: "ら", row: ROWS.RA },
  { sound: "ri", character: "り", row: ROWS.RA },
  { sound: "ru", character: "る", row: ROWS.RA },
  { sound: "re", character: "れ", row: ROWS.RA },
  { sound: "ro", character: "ろ", row: ROWS.RA },

  { sound: "ya", character: "や", row: ROWS.YA },
  { sound: "yu", character: "ゆ", row: ROWS.YA },
  { sound: "yo", character: "よ", row: ROWS.YA },

  { sound: "wa", character: "わ", row: ROWS.WA },
  { sound: "wo", character: "を", row: ROWS.WA },
  { sound: "n", character: "ん", row: ROWS.WA },

  { sound: "ga", character: "が", row: ROWS.ALT },
  { sound: "gi", character: "ぎ", row: ROWS.ALT },
  { sound: "gu", character: "ぐ", row: ROWS.ALT },
  { sound: "ge", character: "げ", row: ROWS.ALT },
  { sound: "go", character: "ご", row: ROWS.ALT },
  { sound: "za", character: "ざ", row: ROWS.ALT },
  { sound: "ji", character: "じ", row: ROWS.ALT },
  { sound: "zu", character: "ず", row: ROWS.ALT },
  { sound: "ze", character: "ぜ", row: ROWS.ALT },
  { sound: "zo", character: "ぞ", row: ROWS.ALT },
  { sound: "da", character: "だ", row: ROWS.ALT },
  { sound: "ji", character: "ぢ", row: ROWS.ALT },
  { sound: "zu", character: "づ", row: ROWS.ALT },
  { sound: "de", character: "で", row: ROWS.ALT },
  { sound: "do", character: "ど", row: ROWS.ALT },
  { sound: "ba", character: "ば", row: ROWS.ALT },
  { sound: "bi", character: "び", row: ROWS.ALT },
  { sound: "bu", character: "ぶ", row: ROWS.ALT },
  { sound: "be", character: "べ", row: ROWS.ALT },
  { sound: "bo", character: "ぼ", row: ROWS.ALT },
  { sound: "pa", character: "ぱ", row: ROWS.ALT },
  { sound: "pi", character: "ぴ", row: ROWS.ALT },
  { sound: "pu", character: "ぷ", row: ROWS.ALT },
  { sound: "pe", character: "ぺ", row: ROWS.ALT },
  { sound: "po", character: "ぽ", row: ROWS.ALT },

  { sound: "kya", character: "きゃ", row: ROWS.YOON },
  { sound: "kyu", character: "きゅ", row: ROWS.YOON },
  { sound: "kyo", character: "きょ", row: ROWS.YOON },
  { sound: "sha", character: "しゃ", row: ROWS.YOON },
  { sound: "shu", character: "しゅ", row: ROWS.YOON },
  { sound: "sho", character: "しょ", row: ROWS.YOON },
  { sound: "cha", character: "ちゃ", row: ROWS.YOON },
  { sound: "chu", character: "ちゅ", row: ROWS.YOON },
  { sound: "cho", character: "ちょ", row: ROWS.YOON },
  { sound: "nya", character: "にゃ", row: ROWS.YOON },
  { sound: "nyu", character: "にゅ", row: ROWS.YOON },
  { sound: "nyo", character: "にょ", row: ROWS.YOON },
  { sound: "hya", character: "ひゃ", row: ROWS.YOON },
  { sound: "hyu", character: "ひゅ", row: ROWS.YOON },
  { sound: "hyo", character: "ひょ", row: ROWS.YOON },
  { sound: "mya", character: "みゃ", row: ROWS.YOON },
  { sound: "myu", character: "みゅ", row: ROWS.YOON },
  { sound: "myo", character: "みょ", row: ROWS.YOON },
  { sound: "rya", character: "りゃ", row: ROWS.YOON },
  { sound: "ryu", character: "りゅ", row: ROWS.YOON },
  { sound: "ryo", character: "りょ", row: ROWS.YOON },
  { sound: "gya", character: "ぎゃ", row: ROWS.YOON },
  { sound: "gyu", character: "ぎゅ", row: ROWS.YOON },
  { sound: "gyo", character: "ぎょ", row: ROWS.YOON },
  { sound: "ja", character: "じゃ", row: ROWS.YOON },
  { sound: "ju", character: "じゅ", row: ROWS.YOON },
  { sound: "jo", character: "じょ", row: ROWS.YOON },
  { sound: "bya", character: "びゃ", row: ROWS.YOON },
  { sound: "byu", character: "びゅ", row: ROWS.YOON },
  { sound: "byo", character: "びょ", row: ROWS.YOON },
  { sound: "pya", character: "ぴゃ", row: ROWS.YOON },
  { sound: "pyu", character: "ぴゅ", row: ROWS.YOON },
  { sound: "pyo", character: "ぴょ", row: ROWS.YOON }
];
