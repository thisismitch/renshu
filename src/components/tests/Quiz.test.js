import React from 'react';
import TestRenderer from 'react-test-renderer';
import Quiz from '../Quiz';
import hiragana from '../../lib/hiragana';

it('renders correctly', () => {
  const tree = TestRenderer.create(<Quiz questions={hiragana} />).toJSON();
  expect(tree).toMatchSnapshot();
});
