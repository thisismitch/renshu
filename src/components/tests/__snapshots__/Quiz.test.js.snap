// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="css-dngzx1"
  />,
  <div
    className="css-165msc7"
  >
    <div
      className="css-16ufovt"
    >
      <h1
        className="css-s1vw9v"
      >
        „ÅÇ
      </h1>
      <input
        autoFocus={true}
        className="css-pr63pg"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
  </div>,
]
`;
